### ISSUE TITLE
after deserialization with schema validation: make detailed type information available (for IDE feedback and mypy checking)

### ISSUE THREAD

**Issue by jgehrcke** [conbench/conbench#796]
Part of #605. We have places in code where we perform precise schema validation of user-given data, but then after schema validation we are not having and not propagating type information (type information that the IDE can make use of, type information that mypy can make use of).

Example:

https://github.com/conbench/conbench/blob/aa95b6d52ff6cb6697b7ab8d04c43d6cf3612860/conbench/api/benchmarks.py#L185

After this, `data` as type `Any`. It is being passed around, and in the consumer all we see is this:

![Screenshot from 2023-03-06 09-29-40](https://user-images.githubusercontent.com/265630/223058097-e66fe08c-7377-4343-842a-3e6e7dc02974.png)

The IDE has no information about  `data`. Understanding what `data` is requires going forth and back in code w/o being able to use IDE tooling. In this particular case we have a marshmallow schema that we can base our ideas on -- but in some details that schema does not _obviously_ translate to Python types. Trial and error is then the only way to know for sure.

After deserializing arbitrary user-given JSON input and validating against a precise schema there _should_ be a way to have precise mypy/IDE-enabling type information on the resulting output object.

The goal of this task here is to make that happen. If that works with marshmallow: fine. If that doesn't work with marshmallow; then we have to change tooling.

------

A bit of an emotional epilogue to this specific ticket here:

To me, the ambiguity "what is this object even about, where do I find its structure?" is pretty much everywhere in the Conbench webapp code. It has significantly slowed down touching any code for me. It is impossible to change code like this 'quickly'. That makes development, well. I want to be honest here: frustratingly slow, and also too error-prone. 

Tiny example; in this line of code: what is `history`? The IDE does not know a thing about it. Will this call to `min()` always succeed, or might it blow up because there may be a `None` item in the input sequence `x["data"]`? https://github.com/conbench/conbench/blob/aa95b6d52ff6cb6697b7ab8d04c43d6cf3612860/conbench/app/_plots.py#L439
In fact, I believe there indeed _may_ be a `None` in the input in which case this line of code will raise a `TypeError`.

**Comment by alistaire47** [comment]
Yes, this is exactly the pain I feel whenever I need to change something in this codebase; improving it would help dramatically!

Relatedly, we do have #430 for validating results fully, ideally in a way such that we can reuse the same code client-side. We do have some type validation already happening, but it's not sufficient to guarantee a result can be processed because we also depend on things like `tags["name"]` being populated. Ideally we'd promote such things to explicitly be part of the schema, but at least having all the validation happen in one place would be a massive improvement over initial type validation and then relying on the processing code for "does it work" validation.

**Comment by jgehrcke** [comment]
Thanks for feedback @alistaire47. It's good that we make this more visible then. Just now, one minute ago I was bitten by:

```
    def get_history_plot(self, benchmark, run, i=0):
>       samples = get_history_for_benchmark(benchmark_id=benchmark.id)
E       AttributeError: 'dict' object has no attribute 'id'
```
I assumed that `benchmark` is a SQLAlchemy-managed object, but it's some dictionary in this case. _some_. It's a dictionary-based representation of a benchmark. It pops up here:
https://github.com/conbench/conbench/blob/aa95b6d52ff6cb6697b7ab8d04c43d6cf3612860/conbench/app/benchmarks.py#L151

But where does it really come from? From here, seemingly:
https://github.com/conbench/conbench/blob/aa95b6d52ff6cb6697b7ab8d04c43d6cf3612860/conbench/app/benchmarks.py#L44

Still not the answer. Following the code further we see that it is the JSON-deserialized version of the benchmark entity.

We're not being dramatic here, this happens too many times per day, and the emotional burden is real.
